cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -541,6 +541,23 @@ class DatasetWriter::DatasetWriterImpl {
     }
   }
 
+  void ResumeIfNeeded() {
+    if (!paused_) {
+      return;
+    }
+    bool needs_resume = false;
+    {
+      std::lock_guard lg(mutex_);
+      if (!write_tasks_ || write_tasks_->QueueSize() == 0) {
+        needs_resume = true;
+      }
+    }
+    if (needs_resume) {
+      paused_ = false;
```

Comment:
Should this be done with the mutex acquired or are all accesses to `paused_` done from the same thread?

Reply:
All access to `paused_` is done from a single "logical thread".  `write_tasks_` is a scheduler with max capacity of 1 and so the items submitted to it will never run in parallel (though they may be run on different OS threads).